proofs/pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,797 @@
+\note{Once the subdomains have been implemented in OpenDP, it will no longer be necessary to list all of the subdomains in the compatible domains section.}
+\begin{definition}[\texttt{AbsoluteDistance(T)}]
+\metricDefn{absolute distance}
+\begin{itemize}
+    \item \textbf{Domain:} \texttt{AllDomain(T)}, where \texttt{T} has the traits \texttt{Sub(Output=T)} (defined in definition \ref{defn:trait-sub}) and \texttt{TotalOrd} (defined in definition \ref{defn:trait-totalord}).
+    \item \textbf{Associated type:} \texttt{Q}.
+    \item \textbf{$d$-close:} For any two elements $n, m$ in \texttt{AllDomain(T)}, where \texttt{T} denotes an arbitrary type with trait \texttt{Sub(Output=T)}, and $d$ of generic type \texttt{Q}, we say that $n, m$ are $d$-close under the absolute distance metric (abbreviated as $d_{Abs}$) whenever
+    \[
+        d_{Abs}(n, m) = |n-m| \leq d.
+    \]
+\end{itemize}
+\end{definition}
+\begin{definition}[\texttt{SymmetricDistance}]
+\metricDefn{symmetric distance} 
+\begin{itemize}
+    \item \textbf{Domains:} \texttt{VectorDomain(inner\_domain)} and \texttt{SizedDomain(VectorDomain(inner\_domain))}, where \texttt{inner\_domain} is any domain.
+    
+    \item \textbf{Associated type:} \texttt{IntDistance}.
+    \item \textbf{$d$-close:} For any two vectors $u, v \in \texttt{VectorDomain(D)}$ and any $d$ of type \texttt{u32}, we say that $u, v$ are $d$-close under the symmetric distance metric (abbreviated as $d_{Sym}$) whenever 
+    \[
+        d_{Sym}(u, v) = |\MultiSet(u) \Delta \MultiSet(v)| \leq d.
+    \]
+\end{itemize}
+\end{definition}
+\emph{Note: the associated type of \texttt{SymmetricDistance} is hard-coded as \texttt{IntDistance}, so when declaring that the metric being used is \texttt{SymmetricDistance}, we only need to write \texttt{metric = SymmetricDistance()}; by contrast, we need to write \texttt{AbsoluteDistance(T)} where \texttt{T} is the type on which we are taking the absolute distance since the associated type for \texttt{AbsoluteDistance} is not hard-coded.}
+
+\begin{definition}[ChangeOneDistance]
+\metricDefn{change-one distance}
+
+\begin{itemize}
+    \item \textbf{Domain:} \texttt{AllDomain(T)}
+    \item \textbf{Associated type:} \texttt{Q}.
+    \item \textbf{$d$-close:} For any two elements $n, m$ in \texttt{AllDomain(T)}, where \texttt{T} denotes an arbitrary type, and $d$ of generic type \texttt{Q}, we say that $n, m$ are $d$-close under the change-one distance metric (abbreviated as $d_{CO}$) whenever
+    \[
+    d_{CO} = |n - m | \leq d \leq 1
+    \]
+\end{itemize}
+
+
+
+\questionr{Are the domain and associated type correct? Are any further traits needed? Is this definition correct?} 
+
+\end{definition}
+
+
+\begin{definition}[DiscreteMetric]
+\metricDefn{discrete metric}
+\begin{itemize}
+    \item \textbf{Domain:} \texttt{AllDomain(T)}
```

Comment:
Why not allow DiscreteMetric on every domain (not just AllDomains)?

Reply:
Maybe this should be DiscreteMetric<T> or DiscreteMetric<DI> if we allow it to be on every domain?